lib/trie/database.go

```diff
@@ -310,94 +339,117 @@ func (t *Trie) WriteDirty(db chaindb.Database) error {
 	return batch.Flush()
 }
 
-func (t *Trie) writeDirty(db chaindb.Batch, curr Node) error {
-	if curr == nil || !curr.IsDirty() {
+func (t *Trie) writeDirty(db chaindb.Batch, n Node) error {
+	if n == nil || !n.IsDirty() {
 		return nil
 	}
 
-	enc, hash, err := curr.EncodeAndHash()
+	encoding, hash, err := n.EncodeAndHash()
 	if err != nil {
-		return err
+		return fmt.Errorf(
+			"cannot encode and hash node with hash 0x%x: %w",
+			n.GetHash(), err)
 	}
 
-	// always hash root even if encoding is under 32 bytes
-	if curr == t.root {
-		h, err := common.Blake2bHash(enc)
+	if n == t.root {
+		// hash root node even if its encoding is under 32 bytes
+		encodingDigest, err := common.Blake2bHash(encoding)
 		if err != nil {
-			return err
+			return fmt.Errorf("cannot hash root node encoding: %w", err)
 		}
 
-		hash = h[:]
+		hash = encodingDigest[:]
 	}
 
-	err = db.Put(hash, enc)
+	err = db.Put(hash, encoding)
 	if err != nil {
-		return err
+		return fmt.Errorf(
+			"cannot put encoding of node with hash 0x%x in database: %w",
+			hash, err)
 	}
 
-	if c, ok := curr.(*node.Branch); ok {
-		for _, child := range c.Children {
-			if child == nil {
-				continue
-			}
+	branch, ok := n.(*node.Branch)
+	if !ok {
+		// the node is a leaf
+		n.SetDirty(false)
+		return nil
+	}
 
-			err = t.writeDirty(db, child)
-			if err != nil {
-				return err
-			}
+	for _, child := range branch.Children {
+		if child == nil {
+			continue
+		}
+
+		err = t.writeDirty(db, child)
+		if err != nil {
+			// Note: do not wrap error since it's returned recursively.
+			return err
 		}
 	}
 
-	curr.SetDirty(false)
+	branch.SetDirty(false)
+
 	return nil
 }
 
-// GetInsertedNodeHashes returns the hash of nodes that are inserted into state trie since last snapshot is called
-// Since inserted nodes are newly created we need to compute their hash values.
-func (t *Trie) GetInsertedNodeHashes() ([]common.Hash, error) {
+// GetInsertedNodeHashes returns the hashes of all nodes that were
+// inserted in the state trie since the last snapshot.
+// We need to compute the hash values of each newly inserted node.
+func (t *Trie) GetInsertedNodeHashes() (hashes []common.Hash, err error) {
 	return t.getInsertedNodeHashes(t.root)
 }
 
-func (t *Trie) getInsertedNodeHashes(curr Node) ([]common.Hash, error) {
-	var nodeHashes []common.Hash
-	if curr == nil || !curr.IsDirty() {
+func (t *Trie) getInsertedNodeHashes(n Node) (hashes []common.Hash, err error) {
+	// TODO pass map of hashes or slice as argument to avoid copying
```

Comment:
@qdm12 could I create an issue to address this TODO?

Reply:
Opened #2115 for it (pointing to this PR), it was quicker than writing an issue imo :smile_cat:

Reply:
I'll merge #2115 in development once this one is merged.